text/3735-field-projections.md

```diff
@@ -0,0 +1,1541 @@
+- `&[mut] MaybeUninit<T>` projections
+  - [project-uninit] provides uninit projections via macros at the projection-site uses `unsafe`
+    internally.
+- multiple of the above
+  - [`field-project`] provides projection for `Pin<&[mut] T>` and `&[mut] MaybeUninit<T>` via a macro
+    at the projection-site: the user writes `proj!($var.$field)` to project to `$field`.
+  - [`field-projection`] is an experimental crate that implements general field projections via a
+    proc-macro that hashes the name of the field to create unique types for each field that can then
+    implement traits to make different output types for projections.
+
+[`field-project`]: https://crates.io/crates/field-project
+[`cell-project`]: https://crates.io/crates/cell-project
+[`pin-projections`]: https://crates.io/crates/pin-projections
+[`project-uninit`]: https://crates.io/crates/project-uninit
+[`field-projection`]: https://crates.io/crates/field-projection
+
+## Blog Posts and Discussions
+
+- [Design Meeting Field Projection](https://hackmd.io/@y86-dev/SkkfRkzWh)
+- [Safe Cell field projection in
+  Rust](https://www.abubalay.com/blog/2020/01/05/cell-field-projection)
+- [Field projdection for `Rc` and
+  `Arc`](https://internals.rust-lang.org/t/field-projection-for-rc-and-arc/15827)
+- [Generic Field Projection](https://internals.rust-lang.org/t/generic-field-projection/16204)
+
+Blog posts about pin (projections):
+- [Pinned places](https://without.boats/blog/pinned-places/)
+- [Overwrite trait](https://smallcultfollowing.com/babysteps/series/overwrite-trait/)
+
+## Rust and Other Languages
+
+Rust already has a precedent for compiler-generated types. All functions and closures have a unique,
+unnameable type.
+
+In C++ there are field projections supported on `std::shared_ptr`, it consists of two pointers, one
+pointing to the reference count and the other to the data. Making it possible to project down to a
+field and still take a reference count on the entire struct, keeping also the field alive.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Syntax Bikeshedding
+
+What is the right syntax for the various operations given in this RFC?
+
+Ideally we would have a strong opinion when this feature is implemented. But the decision should
+only be finalized when stabilizing the feature.
+
+### Field Projection Operator
+
+Current favorite: `$base:expr->$field:ident`.
+
+Alternatives:
+- use `~` instead of `->`
+
+### Naming Field Types
+
+Current favorite: `field_of!($Container:ty, $($fields:expr)+ $(,)?)` macro with `offset_of!` syntax.
+
+Alternatives:
+- Introduce a more native syntax on the level of types `$Container:ty->$field:ident` akin to
+  projecting an expression.
+
+### Declaring a Transparent Container Type
+
+Current favorite: [`#[projecting]`](#projecting-attribute) attribute.
+
+Alternatives:
+- extend the `repr` attribute with
+  - `projecting`
+  - `flatten`
+- use `#[flatten]` instead.
+
```

Comment:
```suggestion

## `Field` trait stability

Should we allow user implementations of the `Field` trait and have user-visible internals for it, or should we make it more opaque and sealed to reserve the possibility of supporting enums or similar?

## Generalized enum projection

Is there a generalization of enum projection that allows for runtime-conditional projection, for structures that sometimes-but-not-always have a given field? This could guarantee the type and identity of the field, but require a projection to be validated against a runtime instance of the value before confirming that the field exists and providing the projected field type. This would allow enums, as well as runtime equivalents such as C-style unions with discriminants or similar mechanisms for identifying variants at runtime.

```